recsplit/recsplit.go

```diff
@@ -0,0 +1,155 @@
+/*
+   Copyright 2021 Erigon contributors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package recsplit
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"math/bits"
+
+	"github.com/ledgerwatch/erigon-lib/etl"
+	"github.com/spaolacci/murmur3"
+)
+
+const RecSplitLogPrefix = "recsplit"
+
+// RecSplit is the implementation of Recursive Split algorithm for constructing perfect hash mapping, described in
+// https://arxiv.org/pdf/1910.06416.pdf Emmanuel Esposito, Thomas Mueller Graf, and Sebastiano Vigna.
+// Recsplit: Minimal perfect hashing via recursive splitting. In 2020 Proceedings of the Symposium on Algorithm Engineering and Experiments (ALENEX),
+// pages 175−185. SIAM, 2020.
+type RecSplit struct {
+	keyExpectedCount uint64 // Number of keys in the hash table
+	keyCount         uint64 // Number of keys actually added to the recSplit (to check the match with keyExpectedCount)
+	bucketCount      uint64 // Number of buckets
+	collector        *etl.Collector
+	built            bool     // Flag indicating that the hash function has been built and no more keys can be added
+	currentBucketIdx uint64   // Current bucket being accumulated
+	currentBucket    [][]byte // Keys in the current bucket accumulated before the recsplit is performed for that bucket
+	builder          Builder
+	bucketSizeAcc    []int // Bucket size accumulator
+	bucketPosAcc     []int // Accumulator for position of every bucket in the encoding of the hash function
+}
+
+// NewRecSplit creates a new RecSplit instance with given number of keys and given bucket size
+// Typical bucket size is 100 - 2000, larger bucket sizes result in smaller representations of hash functions, at a cost of slower access
+func NewRecSplit(keyCount, bucketSize int, tmpDir string) *RecSplit {
+	rs := &RecSplit{keyExpectedCount: uint64(keyCount), bucketCount: uint64((keyCount + bucketSize - 1) / bucketSize)}
+	rs.collector = etl.NewCollector(tmpDir, etl.NewSortableBuffer(etl.BufferOptimalSize))
```

Comment:
here must be SortableBuffer or OldestAppearedBuffer?

Reply:
Sortable, because we are assuming that `AddKey` will be called `keyCount` times, for distinct keys. In fact, I am not really checking distinctness yet, that is a good idea to do

Reply:
I have added the check for duplicate keys, and confirmed that it should be `SortableBuffer`